NetKet/Machine/ffnn.hpp

```diff
@@ -0,0 +1,267 @@
+// Copyright 2018 The Simons Foundation, Inc. - All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Eigen/Dense>
+#include <iostream>
+#include <sstream>
+#include <vector>
+#include "Lookup/lookup.hpp"
+#include "Utils/all_utils.hpp"
+#include "layer.hpp"
+
+#ifndef NETKET_FFNN_HPP
+#define NETKET_FFNN_HPP
+
+namespace netket {
+
+template <typename T>
+class FFNN : public AbstractMachine<T> {
+  using VectorType = typename AbstractMachine<T>::VectorType;
+  using MatrixType = typename AbstractMachine<T>::MatrixType;
+  using Ptype = std::unique_ptr<AbstractLayer<T>>;
+
+  std::vector<Ptype> layers_;  // Pointers to hidden layers
+
+  std::vector<int> layersizes_;
+  int depth_;
+  int nlayer_;
+  int npar_;
+  int nv_;
+
+  const Hilbert &hilbert_;
+
+  const Graph &graph_;
+
+ public:
+  using StateType = typename AbstractMachine<T>::StateType;
+  using LookupType = typename AbstractMachine<T>::LookupType;
+
+  // constructor
+  explicit FFNN(const Graph &graph, const Hilbert &hilbert, const json &pars)
+      : nv_(hilbert.Size()), hilbert_(hilbert), graph_(graph) {
+    Init(pars);
+  }
+
+  void Init(const json &pars) {
+    json layers_par;
+    if (FieldExists(pars["Machine"], "Layers")) {
+      layers_par = pars["Machine"]["Layers"];
+      nlayer_ = layers_par.size();
+    } else {
+      throw InvalidInputError("Field (Layers) not defined for Machine (FFNN)");
+    }
+
+    std::string buffer = "";
```

Comment:
What is the purpose of buffer here? As far as I can see it is always `""` and only passed as the `comment` argument to `InfoMessage`.

Reply:
Yes because I didnt want the "#  "  in the standard messages